### ISSUE TITLE
worker template process

### ISSUE THREAD

**Issue by trgill** [sid-project/sid#108]
Of course, relying on the fact that we're not unwinding the call stack in the worker is far from ideal and it's definitely a pain spot with possible side effects, yes.

This one's a bit tricky though - we can't exit event loop in the parent process (we'd lose events/requests) and we can't exit event loop in the worker (at least not with `sd-event` loop which checks whether the PID hasn't changed on almost all API calls). 

For a clean solution here, we'd either need to change underlying event loop to something else than `sd-event` that supports cleaning up after forking (maybe `libev`?) OR we'd need to create a minimal template process for creating new workers.

(There's also a 3rd option - create our own event loop implementation where we won't be restricted, but we'd need to gain all the know-how that was put into the existing implementations.)

With the "worker template process", the main process would pass the request to this template process and the template process would then fork itself in turn, creating a new worker (without any call stack remnants). The template process would use a very simple event loop to accept these requests ("simple" meaning just to accept those specific requests and easy to exit in forked process).

I think the approach with the "worker template process" is more appropriate as with that, we're not restricting the event loop library we could possibly use. Also, starting a new worker process from a clean template seems much more straightforward. Disadvantage is that it adds intermediary between the main process and worker process, but only during creation. When the workers are already created, the communication can be conducted directly between the main process and that worker process.

_Originally posted by @prajnoha in https://github.com/sid-project/sid/issues/99#issuecomment-859083821_

**Comment by trgill** [comment]
@prajnoha just want to check to make sure I understand the interaction when a udev event arrives.

1. udev event (add|change|remove)  triggers a "usid scan" due to "udev rule"
2. usid scan triggers sd_io_event_handler and calls _on_ubridge_udev_monitor_event
2.a In _on_ubridge_udev_monitor_event we fork creating the child.
3. parent writes the file descriptor to the worker socket for the child to write a response
4. the worker will send 
4.a. WORKER_CHANNEL_CMD_YIELD back if it was just a request to dump the database
4.b. WORKER_CHANNEL_CMD_DATA_EXT if it is a database update that is merged via _worker_proxy_recv_fn()

Questions:

In the case of 4.a we need a current copy of the database in the child, right?  How are you thinking we should deal with this case?
SID gets events from both usid and udev_monitor_new_from_netlink()?  What is the advantage of both?

Thanks!

**Comment by prajnoha** [comment]
> @prajnoha just want to check to make sure I understand the interaction when a udev event arrives.
> 
> 1. udev event (add|change|remove)  triggers a "usid scan" due to "udev rule" or the monitor we setup with udev_monitor_new_from_netlink

Actually, the sequence is this:

  - (udevd listening on "kernel" udev events - the ones coming from the kernel directly)
  - kernel udev event fires
  - udevd starts processing the event, the rules... one of them is `usid scan`
  - `usid scan` communicates with SID daemon through the `sid-ubridge.socket` (a unix domain socket in abstract namespace, so not placed in the filesystem)...
  
> 2. usid scan triggers sd_io_event_handler and calls _on_ubridge_udev_monitor_event

  - ...the `sid-ubridge.socket` is backed by `_on_ubridge_interface_event` callback which gets called...

>    2.a In _on_ubridge_udev_monitor_event we fork creating the child.

  - ...yup, but not the `_on_ubridge_udev_monitor_event`, but the `_on_ubridge_interface_event`...

> 3. parent writes the file descriptor to the worker socket for the child to write a response

  - ...well, yes, but also receiving the request from the "connection socket" - the main process accepts the connection and sends the FD representing the connection to the worker through the "main process <--> worker" channel to handle the actual request sent through that connection/FD. So we have two sockets here to differentiate - one is the socket representing the channel between the main process and worker where main process and worker can interchange data. The other socket is the accepted connection from client. The accepted connection's FD is sent through the main <--> worker channel.

  - on worker side, the `_on_worker_channel_event` callback gets called which in turn calls the `_worker_recv_fn` which then does:
    - creates `sid_resource_type_ubridge_connection` instance to represent the connection, the connection instance has `_on_connection_event` which is called whenever there's any data/request to process on the connection (the socket)
    - when the `_on_connection_event` detects it has full message received, it creates `sid_resource_type_ubridge_command` instance to handle concrete command...
    - the `sid_resource_type_ubridge_command` instance has `_cmd_handler` registered to process the command itself (this is registered as "deferred" event source to not stall the event loop in the worker in case worker would process more commands per connection in the future)
    - the `_cmd_handler` does all the processing in the worker - calls all the phases with modules and their hooks...

> 4. the worker will send
>    4.a. WORKER_CHANNEL_CMD_YIELD back if it was just a request to dump the database
>    4.b. WORKER_CHANNEL_CMD_DATA_EXT if it is a database update that is merged via _worker_proxy_recv_fn()
> 

The `WORKER_CHANNEL_CMD_...` are the commands sent through the internal channel between main process and worker (not between the worker and any external client). And the sequence here is:

  - ...at the end of `_cmd_handler` processing (when all the results are collected), the function sends results:
    - sending results back to the client (the `ucmd_ctx->res_buf` content - the "response buffer)
    - sending results back to the client by passing/sharing the FD representing the buffer memory, currently the database is sent this way:
      -  either to the client - the full database content (for the `sidctl dump` command)
      - or to the main process - only the differences - what has been added/removed/changed in the db snapshot so the main process can merge the diffs (this is the `WORKER_CHANNEL_CMD_DATA_EXT` internal command - we're sending back the extended data - in this case the extended data is the "FD representing the buffer memory" with the db diffs)

When the command is processed, the worker has nothing else to do and so it yields its resources by calling `worker_control_worker_yield` (which in turn issues the `WORKER_CHANNEL_CMD_YIELD` internal command through the "main process <--> worker" channel).

However, this is designed in a way that we can have more commands being processed per one connection. For example, we can call `usid scan`, then we can call `usid checkpoint` from udev rules (several times), so we can "checkpoint" udev rule processing. When udevd finished its own udev rule processing and everything it needs to do for the event processing, it sends out "udev" udev event. And this is what is being monitored in SID with `_on_ubridge_udev_monitor_event`.

So the plan here is to support this sequence:
  - "udev" uevent fires
  - udevd processes udev rules
  - as one of the udev rules, udevd calls `usid scan`
  - SID and its worker does the processing, udevd waits
  - SID and its worker finishes the processing, sends results back to udev
  - more udev rules are processed by udevd
  - some of those rules could be `usid checkpoint`
  - udevd finished processing udev rules and related things and fires the "udev" uevent
  - SID receives the "udev" uevent to confirm that the uevent processing is fully completed
  - now, SID worker can yield its resources (with the `WORKER_CHANNEL_CMD_YIELD` internal command)
  - (the worker might be reused though for further "trigger-action" part processing which comes right after the "udev" uevent is received and hence we are sure that udev has finished its processing for this uevent)

To keep the connection alive for more than one command, we need `usid scan` to be built into udev, which is our plan for the future. Currently, we have our own `usid scan` which is a separate binary, but the disadvantage is that it can't keep the connection alive, because it needs to exit and so the connection is dropped (otherwise it would stall all the other udev rule processing which we can't do).

> Questions:
> 
> 1. In the case of 4.a we need a current copy of the database in the child, right?  How are you thinking we should deal with this case?  Should we move the response to database reads to the parent?  That could work, right?

...the current copy of the database is always acquired by having its snapshot - and that is achieved by forking the main process (creating the worker) in which case we have the COW copy of the in-memory database. So for each command, we do a fork and then we get the db snapshot with it.

Of course, if we had the db outside of memory and we're not relying on the `fork` call itself, we'd need to achieve the snapshot capability by other means, e.g. using `mmap` on the file representing the DB... just an example, depends on what we decide to use in the end...

> 2. SID gets events from both usid and udev_monitor_new_from_netlink()?  What is the advantage of both?

...so as described above - `usid` passes the "kernel" uevent (in the middle of udev processing) to SID while with `udev_monitor_new_from_netlink` we monitor the "udev uevent (after udevd has finished its processing). So we use the first one for our processing on the uevent and responding back to udevd. And the latter one is used for confirmation that udevd has finished its processing and also to start the other stage - the trigger-action part of SID processing.

**Comment by trgill** [comment]
> For a clean solution here, we'd either need to change underlying event loop to something else than `sd-event` that supports cleaning up after forking (maybe `libev`?) OR we'd need to create a minimal template process for creating new workers.
> 
> (There's also a 3rd option - create our own event loop implementation where we won't be restricted, but we'd need to gain all the know-how that was put into the existing implementations.)
> 
> With the "worker template process", the main process would pass the request to this template process and the template process would then fork itself in turn, creating a new worker (without any call stack remnants). The template process would use a very simple event loop to accept these requests ("simple" meaning just to accept those specific requests and easy to exit in forked process).
> 
> I think the approach with the "worker template process" is more appropriate as with that, we're not restricting the event loop library we could possibly use. Also, starting a new worker process from a clean template seems much more straightforward. Disadvantage is that it adds intermediary between the main process and worker process, but only during creation. When the workers are already created, the communication can be conducted directly between the main process and that worker process.
> 
> _Originally posted by @prajnoha in [#99 (comment)](https://github.com/sid-project/sid/pull/99#issuecomment-859083821)_

Thanks for the detailed response!  I understand it, but I'm not sure I see how it connects to the  "worker template process" described above.

My question about needing a current copy of the database was assuming that "starting a new worker process from a clean template" would mean we were starting a new process (not forking with the COW snapshot like we do now).  I think I'm missing something important. :-)

**Comment by prajnoha** [comment]
> Thanks for the detailed response! I understand it, but I'm not sure I see how it connects to the "worker template process" described above.

Ah, I thought you needed to make sure about how *current* code works (...so we can keep the db snapshot functionality, but planning to achieve that with different means than the COW after fork...)

> My question about needing a current copy of the database was assuming that "starting a new worker process from a clean template" would mean we were starting a new process (not forking with the COW snapshot like we do now). I think I'm missing something important. :-)

...so in case we have a template process from which we fork new workers, the important question is how to achieve the snapshot functionality. We need the database access throughout the whole worker processing, because we simply don't know in advance in which phase of the worker processing and how much of the database content the worker needs. With the snapshot, the worker can do this.

So to your question:
> 4. the worker will send
>    4.a. WORKER_CHANNEL_CMD_YIELD back if it was just a request to dump the database
>    4.b. WORKER_CHANNEL_CMD_DATA_EXT if it is a database update that is merged via _worker_proxy_recv_fn()
> 
> Questions:
> 
> 1. In the case of 4.a we need a current copy of the database in the child, right?  How are you thinking we should deal with this case?  Should we move the response to database reads to the parent?  That could work, right?

Well, we always need current copy of the database (snapshot). (...not counting `usid version`, `usid active`, `sidctl tree` commands which do not need to work with the db).

...the `dump` is simple command that doesn't change the db - it just reads the content at one instance (and it's "debug" mostly). For the `scan` command, we need to be able to be able to read consistent db at any of the processing phase in the worker. And the only ways I can think of how to achieve that is either by brute db locking or creating a snapshot. I'd like to avoid that locking as much as possible, because I don't want one worker process to pause all the others (e.g. when we're processing more uevents/scans at once). So that's why I'd like to keep the snapshot...

Yes, currently we rely on the `fork` to do that snapshot for us automatically which goes against the ide of "clean worker process template" created at the beginning from which other worker processes are then created. So this part (creating a db snapshot) needs a different solution first - and instead of relying on the fork itself to do a COW snapshot, we could make use of `mmap` (with `MAP_PRIVATE`) in the worker process that would map the db from the main process (something that comes to my mind, there might be other solutions of course...). I think `lmdb` uses `mmap` to achieve something similar too...

So it all boils down to "how to get snapshot functionality without relying on fork"... before we start using the template process for creating workers.

**Comment by trgill** [comment]
Ok - thanks.  (sorry for the poorly worded questions)

I'll have a look at mmap and lmdb.

How are we going to be sure the workers complete in the order started?  In the case of a fiber channel porting going up/down/up  - we could get 1. add 2. remove 3. add.  So the order of completion for the workers would matter.

**Comment by prajnoha** [comment]
> How are we going to be sure the workers complete in the order started? In the case of a fiber channel porting going up/down/up - we could get 1. add 2. remove 3. add. So the order of completion for the workers would matter.

The "1. add 2.remove" sequence should be already covered by udev itself which doesn't process another uevent for a device where previous uevent for the same device is not yet finished. So if the uevent comes, udev simply takes the `DEVPATH` uevent variable (which is canonical `sysfs` path for a device) and checks if there's any match - either full match or substring match to cover some basic dependent devices like partitions. If there's a match, udev queues the uevent processing, including the udev rules and so including our `usid scan` call (that would create a worker on SID side).

The "2. remove 3. add" sequence can be a bit more tricky though - if the `DEVPATH` is not same across this "remove --> readd" sequence (which may happen), we need more logic to check whether there's any remaining processing for this device to be completed. SID could possibly check its own generated UUID in its own db instead of udev's `DEVPATH` (#107 ... which I recall now that I haven't filled in with info yet!!! Will add, sorry for that missing info!). In addition to that, we have uevent sequence numbers that we can add in the check.

If the `DEVPATH` gets different, udev simply doesn't care and considers the uevent to be for a completely new device. So this is where SID needs to do its own check...

**Comment by prajnoha** [comment]
Regarding the issue with inherited callstack in the worker process after `fork` - I wonder if `clone` call can be of any help here - with that one, it seems we should be able to control more what is inherited from parent process. But this is just a guess, I haven't used or studied `clone` in detail at all. But maybe it's worth looking at whether there's anything in this area that might help us...

**Comment by bmarzins** [comment]
@trgill, is cleaning up the work process after fork something you're looking into, or can I pick this up?

**Comment by trgill** [comment]
@bmarzins please go ahead.  Hopefully you'll have better ideas than I do.  I looked at setting a global - then checking it after the main loop exited - but that seemed a bit awkward.
